### DISCUSSION TITLE
Destroying Body throws Memory Out of Bounds

### DISCUSSION THREAD

**Discussion by DennisSmolek** [jrouwe/JoltPhysics.js#149] [Q&A] [answered]
Working on removing/cleaning up bodies in a simulation. 
I check they exist, remove them, then try to destroy them:
```js
if (!this.bodyInterface.IsAdded(bodyID)) {
            console.log('body already removed');
            return;
        }

        // remove the body from the simulation
        this.bodyInterface.RemoveBody(bodyID);
        // destroy it
        this.bodyInterface.DestroyBody(bodyID);
```

Most of the time this works, however if I do it a few times (3-4) I get a memory error when destroying:
![Screenshot 2024-04-11 183849](https://github.com/jrouwe/JoltPhysics.js/assets/1397052/720f29e1-7f1d-435a-8f28-67925f46dfb7)

I tried switching to just `Jolt.destroy(body)`

and the same error happens:
![Screenshot 2024-04-11 183307](https://github.com/jrouwe/JoltPhysics.js/assets/1397052/b7defc00-534c-487b-9dab-7de0cc1e0c44)

I'm not sure if it's a real memory error in the destroy process of if there's just a lot going on at that moment and that's what finally triggers a memory error. 

Any insights would be appreciated.

**Comment by LeXXik** [comment]
The method looks fine, and it should work correctly, as this example works fine for me:
https://jrouwe.github.io/JoltPhysics.js/add_remove_bodies.html

Which does:
https://github.com/jrouwe/JoltPhysics.js/blob/0a02ddb2a74eafedef64495802bd4994cf7f1a8f/Examples/js/example.js#L202-L205

Which then leaves it to `bodyID` pointing to an already destroyed body. I would look into when you are calling this, e.g. don't do it during some collision event callback.

**Comment by DennisSmolek** [comment]
I think this might be  a red herring error message.
Either something else is causing a general bug or there is some other memory issue, and it only fires on deleting the body.

I have a few others that I'm triggering/testing the same way (reloading within a react router) and I can get them to fire. For example a new Raycast. Which I have another demo that does 60,000 a second no problem. 

Looking in Chrome Devtools I don't see anything obvious and the memory tab doesn't show anything. 
I built with memory profiler but I'm not really sure how to set it up. When the page crashes I get a report page, but that says I'm only using 15% or so of memory.

**Reply by LeXXik** [reply]
Well, from your logs, it looks like the callstack starts from "removeBody" and then proceeds with "destroy". Is it this line?:
```js
bodyInterface.RemoveBody(id); 
```

**Reply by DennisSmolek** [reply]
> Well, from your logs, it looks like the callstack starts from "removeBody" and then proceeds with "destroy". Is it this line?:
> 
> ```js
> bodyInterface.RemoveBody(id); 
> ```

That particular log is from this snippet in the function named `removeBody()`: 
```js
if (!this.bodyInterface.IsAdded(bodyID)) {
    console.log('body already removed');
    return;
}

// remove the body from the simulation
this.bodyInterface.RemoveBody(bodyID);
// destroy it
///this.bodyInterface.DestroyBody(bodyID);
Jolt.destroy(body);
```

With destroy being the standard `Jolt.destroy()` method. 
If I use `bodyInterface.DestroyBody()` it has a slightly longer call stack, but also errors.

**Reply by LeXXik** [reply]
Perhaps you could create a simple repro? 

For example, codesandbox:
https://codesandbox.io/p/devbox/jolt-ygl3sq

Or just modify a local example.

**Reply by DennisSmolek** [reply]
I’ll try tomorrow. The difficult thing is we are making a react wrapper so it’s got some mess in the way.
I’ll try to get it to do the same in vanilla and if I can’t I’ll just share a react repo

**Reply by jrouwe** [reply, marked as answer]
`Jolt.destroy(body)` should crash as `body` was never allocated in the first place (it's part of an internal structure). The only valid way of destroying a body is `bodyInterface.DestroyBody`.  My guess is that you're doing a double destroy, or perhaps you're corrupting memory in a different way. Have you tried running a Debug build to check for asserts that are triggering?

**Reply by DennisSmolek** [reply]
I’m running both debug and memory profiling (although I’m not sure how to utilize that correctly)

in my current setup I’m using the suggested `bodyInterface.DestroyBody()`

That’s what I thought though.. it’s not actually out of bounds but something else is wrong, good to know thats possible

**Reply by jrouwe** [reply]
There's a little bit of css that was needed to present the memory profiler properly, see 1d4070427cad011ef27b8d0bc91a32daf099955d (although I don't think it is strictly needed, without it the 3D viewport became invisible). Another thing to be mindful of is that if you run `./build.sh Debug` followed by `./build-sh Debug -DENABLE_MEMORY_PROFILER=ON` the extra option is not used as the build files are only written the first time you run (this is how CMake works). Remove the `Build` folder to ensure everything gets compiled. It then looks like:

![image](https://github.com/jrouwe/JoltPhysics.js/assets/1621693/cffe3f5c-da48-4d3b-8e76-be9246cc3ac7)
